browser/page_mapping.go

```diff
@@ -206,18 +207,57 @@ func mapPage(vu moduleVU, p *common.Page) mapping { //nolint:gocognit,cyclop
 		"on": func(event string, handler sobek.Callable) error {
 			tq := vu.taskQueueRegistry.get(vu.Context(), p.TargetID())
 
-			mapMsgAndHandleEvent := func(m *common.ConsoleMessage) error {
-				mapping := mapConsoleMessage(vu, m)
-				_, err := handler(sobek.Undefined(), vu.Runtime().ToValue(mapping))
-				return err
-			}
-			runInTaskQueue := func(m *common.ConsoleMessage) {
-				tq.Queue(func() error {
-					if err := mapMsgAndHandleEvent(m); err != nil {
-						return fmt.Errorf("executing page.on handler: %w", err)
+			var runInTaskQueue func(any)
+			switch event {
+			case common.EventPageConsoleAPICalled:
+				mapMsgAndHandleEvent := func(m *common.ConsoleMessage) error {
+					mapping := mapConsoleMessage(vu, m)
+					_, err := handler(sobek.Undefined(), vu.VU.Runtime().ToValue(mapping))
+					return err
+				}
+				runInTaskQueue = func(a any) {
+					tq.Queue(func() error {
+						m, ok := a.(*common.ConsoleMessage)
+						if !ok {
+							return errors.New("incorrect console message")
+						}
+
+						if err := mapMsgAndHandleEvent(m); err != nil {
+							return fmt.Errorf("executing page.on handler: %w", err)
+						}
+						return nil
+					})
+				}
+			case common.EventPageMetricCalled:
+				mapMsgAndHandleEvent := func(m *common.ExportedMetric) error {
+					mapping, err := mapMetric(vu, m)
+					if err != nil {
+						return err
 					}
-					return nil
-				})
+					_, err = handler(sobek.Undefined(), vu.VU.Runtime().ToValue(mapping))
+					return err
+				}
+				runInTaskQueue = func(a any) {
+					tq.Queue(func() error {
+						m, ok := a.(*common.ExportedMetric)
+						if !ok {
+							return errors.New("incorrect metric message")
+						}
+						// page.on('metric') needs to be synchronized so that
+						// the name can be returned back to the caller of the
+						// handler. Once the handler has complete we call
+						// Completed so that the caller knows that the handler
+						// has complete and to work with the new metric.
+						defer m.Completed()
+
+						if err := mapMsgAndHandleEvent(m); err != nil {
+							return fmt.Errorf("executing page.on handler: %w", err)
```

Comment:
Is it executing or mapping? WDYT?

```suggestion
							return fmt.Errorf("mapping page.on handler: %w", err)
```

Reply:
Right, `mapMsgAndHandleEvent` does the mapping and calling of the handler. I don't quite understand why `mapMsgAndHandleEvent` is its own function. I've removed `mapMsgAndHandleEvent` and now the mapping and call to the handler are done from here, with better errors. Done in fd8aa9778ca160ede9d775bc60c0799fabb49e97

Reply:
👍